<a href="https://colab.research.google.com/github/kmendesDev/project_titanic/blob/Data_transformation_Cassio/data_cleaning_titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Projeto Titanic

## Introdução

👋🛳️ Olá, seja bem-vindo ao Desafio Titanic Kaggle! Esta é a lendária competição de Machine Learning do Titanic - o melhor e primeiro desafio de ML no Kaggle.

Esse projeto foi desenvolvido em colaboração com os Cientistas de Dados:

:iphone: Developers contacts

Kaio Mendes - [![General badge](https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://www.linkedin.com/in/kaio-mendes-bba196150)  -  [![General badge](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://github.com/kmendesDev)

Cassio Viana - [![General badge](https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://www.linkedin.com/in/cvs1987)  -  [![General badge](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://github.com/cvs2010)


O objetivo desse projeto é utilizar aprendizado de máquina para criar um modelo que preveja quais passageiros sobreviveram ao naufrágio do Titanic.

O Desafio: O naufrágio do Titanic é um dos naufrágios mais infames da história. Em 15 de abril de 1912, durante sua viagem inaugural, o RMS Titanic, amplamente considerado "inafundável", afundou após colidir com um iceberg. Infelizmente, não havia botes salva-vidas suficientes para todos a bordo, resultando na morte de 1502 dos 2224 passageiros e tripulantes.

Embora tenha havido algum elemento de sorte envolvido na sobrevivência, parece que alguns grupos de pessoas tinham mais probabilidade de sobreviver do que outros.

Neste projeto, iremos desenvolver um modelo preditivo de Machine Learning que responda à pergunta: "que tipo de pessoas tinham mais probabilidade de sobreviver?" usando dados dos passageiros (como nome, idade, gênero, classe socioeconômica, etc).

Quais Dados foram usados?
Dois conjuntos de dados semelhantes que incluem informações dos passageiros, como nome, idade, gênero, classe socioeconômica, etc. Um conjunto de dados é intitulado train.csv e o outro é intitulado test.csv.

O train.csv contém os detalhes de um subconjunto dos passageiros a bordo (exatamente 891) e, o que é mais importante, revelará se eles sobreviveram ou não, também conhecido como "verdade básica". Já o conjunto de dados test.csv contém informações semelhantes, mas não revela a "verdade básica" para cada passageiro. Sua tarefa é prever esses resultados.

Usando os padrões encontrados nos dados do train.csv, prevemos se os 418 passageiros a bordo (encontrados no test.csv) sobreviveram.

Confira a pasta "data" para explorar ainda mais os conjuntos de dados.

https://www.kaggle.com/competitions/titanic/overview/description


##Importação de bibliotecas

In [1]:
import pandas as pd # Para manusear dados tabulares
import seaborn as srn # Para gráficos estatísticos
import statistics as sts # Para cálculos matemáticos
import numpy as np # Para operações numéricas
import seaborn as sns # Para visualização de dados estatísticos
import matplotlib.pyplot as plt # Para criação de gráficos

##Importação dos dados

In [2]:
# Baixar arquivos salvos no github
!wget https://github.com/kmendesDev/project_titanic/raw/main/data/train.csv
!wget https://github.com/kmendesDev/project_titanic/raw/main/data/test.csv
dataset_test = pd.read_csv("/content/test.csv")
dataset_train = pd.read_csv("/content/train.csv")

--2023-08-31 18:41:01--  https://github.com/kmendesDev/project_titanic/raw/main/data/train.csv
Resolving github.com (github.com)... 20.27.177.113
Connecting to github.com (github.com)|20.27.177.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/kmendesDev/project_titanic/main/data/train.csv [following]
--2023-08-31 18:41:01--  https://raw.githubusercontent.com/kmendesDev/project_titanic/main/data/train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 61194 (60K) [text/plain]
Saving to: ‘train.csv’

train.csv           100%[===================>]  59.76K  --.-KB/s    in 0.04s   

2023-08-31 18:41:01 (1.40 MB/s) - ‘train.csv’ saved [61194/61194]

--2023-08-31 18:41:02--  https://github.com/kme

In [3]:
dataset_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
dataset_train.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [5]:
dataset_train.shape

(891, 12)

In [6]:
dataset_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [7]:
dataset_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [8]:
dataset_test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [9]:
dataset_test.tail()

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

In [10]:
dataset_test.shape

(418, 11)

In [11]:
dataset_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [12]:
dataset_test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

##Análise exploratória dos dados

### Análise do conjunto de dados de treinamento

In [84]:
#Clonando o DataFrame
training_data = dataset_train
training_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [85]:
# Informa a identificação do passageiro na lista
training_data['PassengerId'].describe()

count    891.000000
mean     446.000000
std      257.353842
min        1.000000
25%      223.500000
50%      446.000000
75%      668.500000
max      891.000000
Name: PassengerId, dtype: float64

In [86]:
# Coluna Survived:
# Informa a quantidade de 0 e 1 indicando com 1 os que sobreviveram e com 0 os que não conseguiram.
training_data.groupby(['Survived']).size()

Survived
0    549
1    342
dtype: int64

In [87]:
# Verificando a quantidade de itens na coluna Survived
training_data['Survived'].describe()

count    891.000000
mean       0.383838
std        0.486592
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: Survived, dtype: float64

In [88]:
# Coluna Pclass:
# Informa a classe na qual o passageiro viajava:
training_data.groupby(['Pclass']).size()

Pclass
1    216
2    184
3    491
dtype: int64

In [89]:
# Conferindo a quantidade.
training_data['Pclass'].describe()

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [90]:
# Conferindo a quantidade.
training_data['Name'].describe()

count                         891
unique                        891
top       Braund, Mr. Owen Harris
freq                            1
Name: Name, dtype: object

In [91]:
# Coluna sex:
# Informa a quantidade de homens e mulheres:
training_data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [92]:
# Conferindo a quantidade.
training_data['Sex'].describe()

count      891
unique       2
top       male
freq       577
Name: Sex, dtype: object

In [93]:
# Coluna Age:
# Verificar a idade dos passageiros:
training_data['Age'].describe()
# A quantidade está inferior à 891. Logo temos problemas com quantitativo nessa coluna

count    891.000000
mean      29.560236
std       13.005010
min        0.420000
25%       22.000000
50%       29.000000
75%       35.000000
max       80.000000
Name: Age, dtype: float64

In [94]:
# Coluna SibSp:
# Informa a quantidade de irmãos e/ou cônjuges à bordo:
training_data['SibSp'].value_counts()

0    608
1    209
2     28
4     18
3     16
8      7
5      5
Name: SibSp, dtype: int64

In [95]:
# Conferindo a quantidade.
training_data['SibSp'].describe()

count    891.000000
mean       0.523008
std        1.102743
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

In [96]:
# Coluna Parch:
# Informa a quantidade de pais e/ou filhos à bordo:
training_data['Parch'].value_counts()

0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: Parch, dtype: int64

In [97]:
# Conferindo a quantidade.
training_data['Parch'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: Parch, dtype: float64

In [98]:
# Coluna Ticket:
# Informa a numeração do Ticket de cada indivíduo:
training_data['Ticket'].value_counts()

347082      7
CA. 2343    7
1601        7
3101295     6
CA 2144     6
           ..
9234        1
19988       1
2693        1
PC 17612    1
370376      1
Name: Ticket, Length: 681, dtype: int64

In [99]:
# Conferindo a quantidade.
training_data['Ticket'].describe()

count        891
unique       681
top       347082
freq           7
Name: Ticket, dtype: object

In [100]:
# Coluna Fare:
# Informa a tarifa paga:
training_data['Fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [101]:
# Coluna Cabin:
# Informa a cabina que o passageiro ocupava:
training_data['Cabin'].value_counts()
# Verifica-se que a cabine mais recorrente é a B96

B96 B98        4
G6             4
C23 C25 C27    4
C22 C26        3
F33            3
              ..
E34            1
C7             1
C54            1
E36            1
C148           1
Name: Cabin, Length: 147, dtype: int64

In [102]:
# Conferindo a quantidade.
training_data['Cabin'].describe()
# Verifica-se que faltam muitos valores. Temos apenas 204 preenchidos

count         204
unique        147
top       B96 B98
freq            4
Name: Cabin, dtype: object

In [103]:
# Coluna Embarked:
# Informa o porto o qual o passageiro embarcou:
training_data['Embarked'].value_counts()
# O Porto de Southampton (S) é o mais recorrente.

S    646
C    168
Q     77
Name: Embarked, dtype: int64

In [104]:
# Conferindo a quantidade.
training_data['Embarked'].describe()

count     891
unique      3
top         S
freq      646
Name: Embarked, dtype: object

Verificação de valores NaN

In [105]:
training_data.isnull().sum()
# Colunas com valores nulos:
# Age
# Cabin
# Embarked

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

Retirando os valores nulos:

In [106]:
#Substituindo os valores NaN pela média:
training_data['Age'].fillna(int(training_data['Age'].describe()[1]),inplace=True)
int(training_data['Age'].describe()[1])

29

In [107]:
#Verificando a existência de algum valor NaN:
training_data['Age'].isnull().sum()

0

In [108]:
# Coluna Cabin:
training_data['Cabin'].isnull().sum()
#Verifica-se 687 valores nulos, caracterizando mais da metade.

687

In [109]:
# Coluna Embarked:
training_data['Embarked'].isnull().sum()
#Verifica-se 2 valores nulos na coluna Embarked.

0

In [110]:
training_data['Embarked'].fillna('S',inplace=True)

In [111]:
# Verificando se ainda há valores nulos:
training_data['Embarked'].isnull().sum()

0

In [112]:
#Verificando a nova quantidade de dados por variável:
training_data['Embarked'].value_counts()

S    646
C    168
Q     77
Name: Embarked, dtype: int64

Verificando valores duplicados de algumas colunas: Id, Name e Cabin

In [113]:
# Coluna PassengerId
training_data[training_data.duplicated(['PassengerId'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [114]:
# Coluna Name
training_data[training_data.duplicated(['Name'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [115]:
# Filtra os dados duplicados e não nulos na coluna 'Cabin'
filtered_data_cabin = training_data[training_data['Cabin'].duplicated(keep=False) & training_data['Cabin'].notna()]
# Ordena os dados
filtered_data_cabin = filtered_data_cabin.sort_values(by='Cabin')
# Excluindo o indice
filtered_data_cabin.reset_index(drop=True)
# Exibe os dados
filtered_data_cabin

PassengerId  Survived  Pclass  \
329          330         1       1   
523          524         1       1   
781          782         1       1   
690          691         1       1   
745          746         0       1   
..           ...       ...     ...   
618          619         1       2   
251          252         0       3   
10            11         1       3   
205          206         0       3   
394          395         1       3   

                                                  Name     Sex   Age  SibSp  \
329                       Hippach, Miss. Jean Gertrude  female  16.0      0   
523    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  female  44.0      0   
781          Dick, Mrs. Albert Adrian (Vera Gillespie)  female  17.0      1   
690                            Dick, Mr. Albert Adrian    male  31.0      1   
745                       Crosby, Capt. Edward Gifford    male  70.0      1   
..                                                 ...     ...   ...    ...   
618                        Becker, Miss. Marion Louise  female   4.0      2   
251         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
205                         Strom, Miss. Telma Matilda  female   2.0      0   
394  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female  24.0      0   

     Parch     Ticket     Fare Cabin Embarked  
329      1     111361  57.9792   B18        C  
523      1     111361  57.9792   B18        C  
781      0      17474  57.0000   B20        S  
690      0      17474  57.0000   B20        S  
745      1  WE/P 5735  71.0000   B22        S  
..     ...        ...      ...   ...      ...  
618      1     230136  39.0000    F4        S  
251      1     347054  10.4625    G6        S  
10       1    PP 9549  16.7000    G6        S  
205      1     347054  10.4625    G6        S  
394      2    PP 9549  16.7000    G6        S  

[103 rows x 12 columns]

Alguns passageiros não tem cabine registradas: 687

Alguns passageiros tem cabines duplicadas: 103

Alguns passageiros tem mais de uma cabine registrada, provavelmente cabines conexas: 16

Passageiros registrados na mesma cabine tiveram diferentes taxas de sobrevivência, isso pode evidenciar algo relacionado ao posicionamento da pessoa no momento do naufrágio, ou outra característica como idade, dificuldade de locomoção, etc.

### Análise do conjunto de dados de teste

In [150]:
#Clonando o DataFrame teste
testing_data = dataset_test
testing_data

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male  30.0      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male  30.0      0      0              359309    8.0500   NaN        S  
417    male  30.0      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [151]:
# Informa a identificação do passageiro na lista
testing_data['PassengerId'].describe()

count     418.000000
mean     1100.500000
std       120.810458
min       892.000000
25%       996.250000
50%      1100.500000
75%      1204.750000
max      1309.000000
Name: PassengerId, dtype: float64

In [152]:
# Coluna Pclass:
# Informa a classe na qual o passageiro viajava:
testing_data.groupby(['Pclass']).size()

Pclass
1    107
2     93
3    218
dtype: int64

In [153]:
# Conferindo a quantidade.
testing_data['Pclass'].describe()

count    418.000000
mean       2.265550
std        0.841838
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [154]:
# Conferindo a quantidade.
testing_data['Name'].describe()

count                  418
unique                 418
top       Kelly, Mr. James
freq                     1
Name: Name, dtype: object

In [155]:
# Coluna sex:
# Informa a quantidade de homens e mulheres:
testing_data['Sex'].value_counts()

male      266
female    152
Name: Sex, dtype: int64

In [156]:
# Conferindo a quantidade.
testing_data['Sex'].describe()

count      418
unique       2
top       male
freq       266
Name: Sex, dtype: object

In [157]:
# Coluna Age:
# Verificar a idade dos passageiros:
testing_data['Age'].describe()
# A quantidade está inferior à 418. Logo temos problemas com quantitativo nessa coluna

count    418.000000
mean      30.216507
std       12.635016
min        0.170000
25%       23.000000
50%       30.000000
75%       35.750000
max       76.000000
Name: Age, dtype: float64

In [158]:
# Coluna SibSp:
# Informa a quantidade de irmãos e/ou cônjuges à bordo:
testing_data['SibSp'].value_counts()

0    283
1    110
2     14
3      4
4      4
8      2
5      1
Name: SibSp, dtype: int64

In [159]:
# Conferindo a quantidade.
testing_data['SibSp'].describe()

count    418.000000
mean       0.447368
std        0.896760
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

In [160]:
# Coluna Parch:
# Informa a quantidade de pais e/ou filhos à bordo:
testing_data['Parch'].value_counts()

0    324
1     52
2     33
3      3
4      2
9      2
6      1
5      1
Name: Parch, dtype: int64

In [161]:
# Conferindo a quantidade.
testing_data['Parch'].describe()

count    418.000000
mean       0.392344
std        0.981429
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        9.000000
Name: Parch, dtype: float64

In [162]:
# Coluna Ticket:
# Informa a numeração do Ticket de cada indivíduo:
testing_data['Ticket'].value_counts()

PC 17608    5
CA. 2343    4
113503      4
PC 17483    3
220845      3
           ..
349226      1
2621        1
4133        1
113780      1
2668        1
Name: Ticket, Length: 363, dtype: int64

In [163]:
# Conferindo a quantidade.
testing_data['Ticket'].describe()

count          418
unique         363
top       PC 17608
freq             5
Name: Ticket, dtype: object

In [164]:
# Coluna Fare:
# Informa a tarifa paga:
testing_data['Fare'].describe()

count    418.000000
mean      35.625688
std       55.840509
min        0.000000
25%        7.895800
50%       14.454200
75%       31.500000
max      512.329200
Name: Fare, dtype: float64

In [165]:
# Coluna Fare:
# Informa a tarifa paga:
testing_data['Fare'].isnull().sum()
# Apenas 1 valor nulo

0

In [166]:
#Substituindo os valores NaN pela média:
testing_data['Fare'].fillna(int(testing_data['Fare'].describe()[1]),inplace=True)
int(testing_data['Fare'].describe()[1])

35

In [167]:
# Coluna Fare:
# Informa a tarifa paga:
testing_data['Fare'].isnull().sum()

0

In [168]:
# Coluna Cabin:
# Informa a cabina que o passageiro ocupava:
testing_data['Cabin'].value_counts()
# Verifica-se que a cabine mais recorrente é a B96

B57 B59 B63 B66    3
B45                2
C89                2
C55 C57            2
A34                2
                  ..
E52                1
D30                1
E31                1
C62 C64            1
C105               1
Name: Cabin, Length: 76, dtype: int64

In [169]:
# Conferindo a quantidade.
testing_data['Cabin'].describe()
# Verifica-se que faltam muitos valores. Temos apenas 204 preenchidos

count                  91
unique                 76
top       B57 B59 B63 B66
freq                    3
Name: Cabin, dtype: object

In [170]:
# Coluna Embarked:
# Informa o porto o qual o passageiro embarcou:
testing_data['Embarked'].value_counts()
# O Porto de Southampton (S) é o mais recorrente.

S    270
C    102
Q     46
Name: Embarked, dtype: int64

In [171]:
# Conferindo a quantidade.
testing_data['Embarked'].describe()

count     418
unique      3
top         S
freq      270
Name: Embarked, dtype: object

Verificação de valores NaN

In [172]:
testing_data.isnull().sum()
# Colunas com valores nulos:
# Age
# Cabin
# Embarked

PassengerId      0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          327
Embarked         0
dtype: int64

Retirando os valores nulos:

In [173]:
#Substituindo os valores NaN pela média:
testing_data['Age'].fillna(int(testing_data['Age'].describe()[1]),inplace=True)
int(testing_data['Age'].describe()[1])

30

In [174]:
#Verificando a existência de algum valor NaN:
testing_data['Age'].isnull().sum()

0

In [175]:
# Coluna Cabin:
testing_data['Cabin'].isnull().sum()
#Verifica-se 327 valores nulos, caracterizando mais da metade.

327

In [176]:
# Coluna Embarked:
testing_data['Embarked'].isnull().sum()

0

Verificando valores duplicados de algumas colunas: Id, Name e Cabin

In [177]:
# Coluna PassengerId
testing_data[testing_data.duplicated(['PassengerId'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [178]:
# Coluna Name
testing_data[testing_data.duplicated(['Name'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [179]:
# Filtra os dados duplicados e não nulos na coluna 'Cabin'
filtered_data_cabin2 = testing_data[testing_data['Cabin'].duplicated(keep=False) & testing_data['Cabin'].notna()]
# Ordena os dados
filtered_data_cabin2 = filtered_data_cabin2.sort_values(by='Cabin')
# Excluindo o indice
filtered_data_cabin2.reset_index(drop=True)
# Exibe os dados
filtered_data_cabin2

PassengerId  Pclass                                               Name  \
293         1185       1                              Dodge, Dr. Washington   
374         1266       1              Dodge, Mrs. Washington (Ruth Vidaver)   
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
287         1179       1                         Snyder, Mr. John Pillsbury   
24           916       1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)   
64           956       1                        Ryerson, Master. John Borie   
142         1034       1                         Ryerson, Mr. Arthur Larned   
356         1248       1     Brown, Mrs. John Murray (Caroline Lane Lamson)   
77           969       1  Cornell, Mrs. Robert Clifford (Malvina Helen L...   
177         1069       1                    Stengel, Mr. Charles Emil Henry   
100          992       1  Stengel, Mrs. Charles Emil Henry (Annie May Mo...   
69           961       1                Fortune, Mrs. Mark (Mary McDougald)   
53           945       1                         Fortune, Miss. Ethel Flora   
50           942       1                           Smith, Mr. Lucien Philip   
395         1287       1     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)   
81           973       1                                 Straus, Mr. Isidor   
114         1006       1             Straus, Mrs. Isidor (Rosalie Ida Blun)   
118         1010       1                               Beattie, Mr. Thomson   
270         1162       1                       McCaffry, Mr. Thomas Francis   
411         1303       1    Minahan, Mrs. William Edward (Lillian E Thorpe)   
34           926       1                           Mock, Mr. Philipp Edmund   
218         1110       1       Widener, Mrs. George Dunton (Eleanor Elkins)   
407         1299       1                         Widener, Mr. George Dunton   
272         1164       1      Clark, Mrs. Walter Miller (Virginia McDowell)   
252         1144       1                           Clark, Mr. Walter Miller   
242         1134       1                       Spedden, Mr. Frederic Oakley   
196         1088       1                    Spedden, Master. Robert Douglas   
326         1218       2                       Becker, Miss. Ruth Elizabeth   
178         1070       2   Becker, Mrs. Allen Oliver (Nellie E Baumgardner)   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
293    male  53.0      1      1     33638   81.8583              A34        S  
374  female  54.0      1      1     33638   81.8583              A34        S  
12   female  23.0      1      0     21228   82.2667              B45        S  
287    male  24.0      1      0     21228   82.2667              B45        S  
24   female  48.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C  
64     male  13.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
142    male  61.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C  
356  female  59.0      2      0     11769   51.4792             C101        S  
77   female  55.0      2      0     11770   25.7000             C101        S  
177    male  54.0      1      0     11778   55.4417             C116        C  
100  female  43.0      1      0     11778   55.4417             C116        C  
69   female  60.0      1      4     19950  263.0000      C23 C25 C27        S  
53   female  28.0      3      2     19950  263.0000      C23 C25 C27        S  
50     male  24.0      1      0     13695   60.0000              C31        S  
395  female  18.0      1      0     13695   60.0000              C31        S  
81     male  67.0      1      0  PC 17483  221.7792          C55 C57        S  
114  female  63.0      1      0  PC 17483  221.7792          C55 C57        S  
118    male  36.0      0      0     13050   75.2417               C6        C  
270    male  46.0      0      0     13050   75.2417               C6        C  
411  female  37.0      1      0   

In [180]:
testing_data['Cabin'].isnull().sum()

327

In [181]:
filtered_data_cabin2.shape

(29, 11)

Alguns passageiros não tem cabine registradas: 327

Alguns passageiros tem cabines duplicadas: 29

Alguns passageiros tem mais de uma cabine registrada, provavelmente cabines conexas: 7

Passageiros registrados na mesma cabine tiveram diferentes taxas de sobrevivência, isso pode evidenciar algo relacionado ao posicionamento da pessoa no momento do naufrágio, ou outra característica como idade, dificuldade de locomoção, etc.

## Pré-processamento, limpeza e tratamento dos dados

### Pré-processamento e limpeza dos dados

In [192]:
testing_data.isnull().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [193]:
training_data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Existem valores ausentessomente na coluna Cabin.

Vamos proceder com o tratamento dos dados ausentes

In [194]:
# Substituindo os valores NaN de Cabin por zero:
training_data['Cabin'].fillna(0,inplace=True)
#Verificando a existência de algum valor NaN:
training_data['Cabin'].isnull().sum()

# Substituindo os valores NaN de Cabin por zero:
testing_data['Cabin'].fillna(0,inplace=True)
#Verificando a existência de algum valor NaN:
testing_data['Cabin'].isnull().sum()

0

In [195]:
training_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500     0        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250     0        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500     0        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000     0        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500     0        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500     0        Q  

[891 rows x 12 columns]

In [196]:
training_data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [197]:
testing_data

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292     0        Q  
1    female  47.0      1      0              363272    7.0000     0        S  
2      male  62.0      0      0              240276    9.6875     0        Q  
3      male  27.0      0      0              315154    8.6625     0        S  
4    female  22.0      1      1             3101298   12.2875     0        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male  30.0      0      0           A.5. 3236    8.0500     0        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500     0        S  
416    male  30.0      0      0              359309    8.0500     0        S  
417    male  30.0      1      1                2668   22.3583     0        C  

[418 rows x 11 columns]

In [198]:
testing_data.isnull().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Confirmamos que não há mais nenhum valor ausentes nos dois datasets

### Tratamento dos dados


Agora iremos para a segunda etapa de modificação das colunas com valores não numéricos. Essa etapa é chamada de "codificação" e visa transformar os valores não numéricos em formatos que possam ser interpretados por algoritmos de análise e modelagem. Isso é essencial para garantir que os dados sejam adequados para análises estatísticas e de aprendizado de máquina, permitindo que os modelos compreendam as informações presentes nas colunas não numéricas. Existem várias técnicas de codificação, como a codificação one-hot e a codificação de rótulos, que podem ser aplicadas dependendo das características dos dados e dos objetivos da análise.

#### Vamos proceder com a codificação dos dados do dataset de treino

In [199]:
# Lista de colunas numéricas
colunas_numericas = training_data.select_dtypes(include=[np.number]).columns.tolist()

# Lista de colunas não numéricas
colunas_nao_numericas = training_data.select_dtypes(exclude=[np.number]).columns.tolist()

print("Colunas Numéricas:")
print(colunas_numericas)

print("\nColunas Não Numéricas:")
print(colunas_nao_numericas)

Colunas Numéricas:
['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']

Colunas Não Numéricas:
['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']


A coluna nome não precisa de nenhum tipo de tratamento pois será desconsiderada para a análise do Machine Learning. Vamos prosseguir para as demais colunas.

In [200]:
# Realizando one-hot encoding na coluna "Sex"
one_hot_encoded = pd.get_dummies(training_data['Sex'], prefix='Sex')

# Adicionando as colunas codificadas ao DataFrame original
training_data = pd.concat([training_data, one_hot_encoded], axis=1)

In [201]:
# Realizando one-hot encoding na coluna "Embarked"
one_hot_encoded = pd.get_dummies(training_data['Embarked'], prefix='Embarked')

# Adicionando as colunas codificadas ao DataFrame original
dataset = pd.concat([training_data, one_hot_encoded], axis=1)

In [203]:
training_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Sex_female  Sex_male  
0        0         A/5 21171   7.2500     0        S           0         1  
1        0          PC 17599  71.2833   C85        C           1         0  
2        0  STON/O2. 3101282   7.9250     0        S           1         0  
3        0            113803  53.1000  C123        S           1         0  
4        0            373450   8.0500     0        S           0         1  
..     ...               ...      ...   ...      ...         ...       ...  
886      0            211536  13.0000     0        S           0         1  
887      0            112053  30.0000   B42        S           1         0  
888      2        W./C. 6607  23.4500     0        S           1         0  
889      0            111369  30.0000  C148        C           0         1  
890      0            370376   7.7500     0        Q           0         1  

[891 rows x 14 columns]

#### Para as colunas "Ticket" e "Cabin" o trabalho será mais árduo, pois teremos que fazer um pré-processamento antes da codificação, devido a grande quantidade de valores diferenciados nessas colunas, veja:

In [204]:
# Verificando os valores únicos em "Ticket" e "Cabin"
print('Valores únicos em Ticket:', len(training_data['Ticket'].value_counts()))
print('Valores únicos em Cabin:', len(training_data['Cabin'].value_counts()))

Valores únicos em Ticket: 681
Valores únicos em Cabin: 148


##### Vamos proceder com a transformação da coluna Ticket

In [227]:
# Criando uma nova coluna "Ticket_value" e outra "Ticket_number"
training_data['Ticket_value'] = training_data['Ticket'].str.split(' ').str[0]
training_data['Ticket_number'] = training_data['Ticket'].str.split(' ').str[1]

# Copiando valores de "Ticket_value" para "Ticket_number" quando estiver com valor NaN
training_data['Ticket_number'].fillna(training_data['Ticket_value'], inplace=True)
for i in range(len(training_data['Ticket_number'])):
  if training_data['Ticket_number'][i].isnumeric() == False:
    training_data['Ticket_number'][i] = 0

# Transformando valores numéricos em NaN na coluna "Ticket_value"
training_data['Ticket_value'] = np.where(pd.to_numeric(training_data['Ticket_value'], errors='coerce').notnull(), np.nan, training_data['Ticket_value'])

# Exibindo as primeiras linhas do DataFrame com a coluna atualizada
training_data

<ipython-input-227-ec07ce1e7cd6>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  training_data['Ticket_number'][i] = 0


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Sex_female  Sex_male  \
0        0         A/5 21171   7.2500     0        S           0         1   
1        0          PC 17599  71.2833   C85        C           1         0   
2        0  STON/O2. 3101282   7.9250     0        S           1         0   
3        0            113803  53.1000  C123        S           1         0   
4        0            373450   8.0500     0        S           0         1   
..     ...               ...      ...   ...      ...         ...       ...   
886      0            211536  13.0000     0        S           0         1   
887      0            112053  30.0000   B42        S           1         0   
888      2        W./C. 6607  23.4500     0        S           1         0   
889      0            111369  30.0000  C148        C           0         1   
890      0            370376   7.7500     0        Q           0         1   

    Ticket_value Ticket_number  Ticket_value_encoded  Ticket_final  
0            A/5         21171                     1      121171.0  
1             PC         17599                     2      217599.0  
2       STON/O2.       3101282                     3    33101282.0  
3            NaN        113803                     4     4113803.0  
4            NaN        373450                     4     4373450.0  
..           ...           ...                   ...           ...  
886          NaN        211536                     4     4211536.0  
887          NaN        112053                     4     4112053.0  
888        W./C.          6607                    16      166607.0  
889          NaN        111369                     4     4111369.0  
890          NaN        370376                     4     4370376.0  

[891 rows x 18 columns]

In [229]:
# Verificando os valores únicos em "Ticket Value"
training_data['Ticket_value'].value_counts()

PC            60
C.A.          27
STON/O        12
A/5           10
W./C.          9
CA.            8
SOTON/O.Q.     8
A/5.           7
SOTON/OQ       7
STON/O2.       6
CA             6
F.C.C.         5
C              5
SC/PARIS       5
S.O.C.         5
SC/Paris       4
LINE           4
A/4            3
PP             3
A/4.           3
S.O./P.P.      3
SC/AH          3
A.5.           2
A./5.          2
P/PP           2
WE/P           2
SOTON/O2       2
S.C./PARIS     2
S.W./PP        1
F.C.           1
A/S            1
SO/C           1
W/C            1
SC             1
SCO/W          1
SW/PP          1
Fa             1
S.O.P.         1
S.C./A.4.      1
A4.            1
S.P.           1
W.E.P.         1
C.A./SOTON     1
Name: Ticket_value, dtype: int64

In [230]:
# Verificando a quantidade de valores distintos
training_data['Ticket_value'].nunique()

43

In [231]:
# Criando um mapeamento de valores únicos para códigos começando de 1 (dicionário)
unique_values = training_data['Ticket_value'].unique()
value_to_code = {value: code+1 for code, value in enumerate(unique_values)}

# Aplicando o mapeamento para criar a nova coluna codificada
training_data['Ticket_value_encoded'] = training_data['Ticket_value'].map(value_to_code).fillna(0).astype(int)

# Exibindo as primeiras linhas do DataFrame com a coluna codificada
training_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  Sex_female  Sex_male  \
0        0         A/5 21171   7.2500     0        S           0         1   
1        0          PC 17599  71.2833   C85        C           1         0   
2        0  STON/O2. 3101282   7.9250     0        S           1         0   
3        0            113803  53.1000  C123        S           1         0   
4        0            373450   8.0500     0        S           0         1   
..     ...               ...      ...   ...      ...         ...       ...   
886      0            211536  13.0000     0        S           0         1   
887      0            112053  30.0000   B42        S           1         0   
888      2        W./C. 6607  23.4500     0        S           1         0   
889      0            111369  30.0000  C148        C           0         1   
890      0            370376   7.7500     0        Q           0         1   

    Ticket_value Ticket_number  Ticket_value_encoded  Ticket_final  
0            A/5         21171                     1      121171.0  
1             PC         17599                     2      217599.0  
2       STON/O2.       3101282                     3    33101282.0  
3            NaN        113803                     4     4113803.0  
4            NaN        373450                     4     4373450.0  
..           ...           ...                   ...           ...  
886          NaN        211536                     4     4211536.0  
887          NaN        112053                     4     4112053.0  
888        W./C.          6607                    16      166607.0  
889          NaN        111369                     4     4111369.0  
890          NaN        370376                     4     4370376.0  

[891 rows x 18 columns]

In [254]:
# Concatenando a coluna "Ticket_value_encoded" com a coluna "Ticket_number" em "Ticket_final"
training_data['Ticket_final'] = training_data['Ticket_value_encoded'].astype(str) + training_data['Ticket_number'].astype(str)

# Transformando a coluna "Ticket_final" em valores numéricos
training_data['Ticket_final'] = pd.to_numeric(training_data['Ticket_final'])

# Exibindo as primeiras linhas do DataFrame com a coluna numérica "Ticket_final"
training_data.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Sex_female  Sex_male  \
0      0         A/5 21171   7.2500     0        S           0         1   
1      0          PC 17599  71.2833   C85        C           1         0   
2      0  STON/O2. 3101282   7.9250     0        S           1         0   

  Ticket_value Ticket_number  Ticket_value_encoded  Ticket_final  
0          A/5         21171                     1        121171  
1           PC         17599                     2        217599  
2     STON/O2.       3101282                     3      33101282

##### Agora vamos transformar a coluna "Cabin" em numérica

In [256]:
training_data['Cabin'].value_counts()

0              687
C23 C25 C27      4
G6               4
B96 B98          4
C22 C26          3
              ... 
E34              1
C7               1
C54              1
E36              1
C148             1
Name: Cabin, Length: 148, dtype: int64

As letras dos códigos das Cabines segundo referência da planta do Titanic do  são: 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'

In [347]:
# Dicionário de mapeamento de letras para números
letter_to_number = {'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E': 5, 'F': 6, 'G': 7, 'T': 8}

# Criando uma nova coluna "Cabin_encoded" usando o dicionário de mapeamento
# Função para mapear letras para números
def map_letters_to_numbers(cabin):
    if isinstance(cabin, str):
        new_cabin = ''
        for letter in cabin:
            if letter in letter_to_number:
                new_cabin += str(letter_to_number[letter])
        return int(new_cabin) if new_cabin else None
    else:
        return None

# Aplicando a função à coluna "Cabin"
training_data['Cabin_encoded'] = training_data['Cabin'].apply(map_letters_to_numbers).astype('Int64')

# Obtendo a parte numérica da coluna "Cabin" de 0 a 9
training_data['Cabin_number'] = training_data['Cabin'].str.extractall('(\d+)').groupby(level=0).apply(lambda x: ''.join(x[0]))

# Combinando "Cabin_encoded" e "Cabin_number" em uma nova coluna "Cabin_combined"
training_data['Cabin_combined'] = training_data['Cabin_encoded'].astype(str) + training_data['Cabin_number'].fillna('')

# Removendo o ponto '.' da coluna "Cabin_combined"
training_data['Cabin_combined'] = training_data['Cabin_combined'].str.replace('.', '', regex=False)

# Substituindo "nan" por 0 na coluna "Cabin_combined"
training_data['Cabin_combined'] = training_data['Cabin_combined'].replace('nan', '0')

# Transformando a coluna "Cabin_combined" em valores numéricos
training_data['Cabin_combined'] = pd.to_numeric(training_data['Cabin_combined'], errors='coerce').astype('Int64')

# Substituindo os valores ausentes em "Cabin_combined" por 0
training_data['Cabin_combined'].fillna(0, inplace=True)

# Exibindo as primeiras linhas do DataFrame com a coluna "Cabin_encoded"
training_data.head(30)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   
20           21         0       2   
21           22         1       2   
22           23         1       3   
23           24         1       1   
24           25         0       3   
25           26         1       3   
26           27         0       3   
27           28         0       1   
28           29         1       3   
29           30         0       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male  29.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male  29.0      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female  29.0      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
22                        McGowan, Miss. Anna "Annie"  female  15.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
24                      Palsson, Miss. Torborg Danira  female   8.0      3   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26                            Emir, Mr. Farred Chehab    male  29.0      0   
27                     Fortune, Mr. Charles Alexander    male  19.0      3   
28                      O'Dwyer, Miss. Ellen "Nellie"  female  29.0      0   
29                                Todoroff, Mr. Lalio    male  29.0      0   

    Parch            Ticket      Fare  ... Embarked Sex_female  Sex_male  \
0       0         A/5 21171    7.2500  ...        S          0         1   
1       0          PC 17599   71.2833  ...        C          1         0   
2       0  STON/O2. 3101282    7.9250  ...        S          1         0   
3       0            113803   53.1000  ...        S          1         0   
4       0            373450    8.0500  ...        S      

### Dataset com limpeza e tratamento finalizado

In [ ]:
dataset_cleaned = dataset
dataset_cleaned

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare  ... Embarked_C Embarked_Q  \
0         0           A/5 21171    7.2500  ...          0          0   
1         0            PC 17599   71.2833  ...          1          0   
2         0    STON/O2. 3101282    7.9250  ...          0          0   
3         0              113803   53.1000  ...          0          0   
4         0              373450    8.0500  ...          0          0   
...     ...                 ...       ...  ...        ...        ...   
1304      0           A.5. 3236    8.0500  ...          0          0   
1305      0            PC 17758  108.9000  ...          1          0   
1306      0  SOTON/O.Q. 3101262    7.2500  ...          0          0   
1307      0              359309    8.0500  ...          0          0   
1308      1                2668   22.3583  ...          1          0   

      Embarked_S  Ticket_value  Ticket_number  Ticket_value_encoded  \
0              1           A/5          21171                     1   
1              0            PC          17599                     2   
2              1      STON/O2.        3101282                     3   
3              1           NaN         113803                     4   
4              1           NaN         373450                     4   
...          ...           ...            ...                   ...   
1304           1          A.5.           3236                    25   
1305           0            PC          17758                     2   
1306           1    SOTON/O.Q.        3101262                    22   
1307           1           NaN         359309                     4   
1308           0           NaN           2668                     4   

      Ticket_final Cabin_encoded Cabin_number  Cabin_combined  
0         211711.0           NaN          NaN               0  
1         175992.0           3.0           85            3085  
2       31012823.0           NaN          NaN               0  
3        1138034.0           3.0          123           30123  
4        3734504.0           NaN          NaN               0  
...            ...           ...          ...             ...  
1304      323625.0           NaN          NaN               0  
1305      177582.0           3.0          105           30105  
1306   310126222.0           NaN          NaN               0  
1307     3593094.0           NaN          NaN               0  
1308       26684.0           NaN          NaN               0  

[1309 rows x 24 columns]

In [ ]:
# Substituindo os valores NaN por zero nas colunas "Cabin_encoded" e "Cabin_number"
dataset_cleaned['Cabin_encoded'].fillna(0, inplace=True)
dataset_cleaned['Cabin_number'].fillna(0, inplace=True)

# Transformando valores das colunas "Ticket_number" e "Cabin_number" em numéricos
dataset_cleaned['Ticket_number'] = pd.to_numeric(dataset_cleaned['Ticket_number'], errors='coerce')
dataset_cleaned['Cabin_number'] = pd.to_numeric(dataset_cleaned['Cabin_number'], errors='coerce')

# Exibindo as primeiras linhas do DataFrame com as colunas atualizadas
dataset_cleaned

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare  ... Embarked_C Embarked_Q  \
0         0           A/5 21171    7.2500  ...          0          0   
1         0            PC 17599   71.2833  ...          1          0   
2         0    STON/O2. 3101282    7.9250  ...          0          0   
3         0              113803   53.1000  ...          0          0   
4         0              373450    8.0500  ...          0          0   
...     ...                 ...       ...  ...        ...        ...   
1304      0           A.5. 3236    8.0500  ...          0          0   
1305      0            PC 17758  108.9000  ...          1          0   
1306      0  SOTON/O.Q. 3101262    7.2500  ...          0          0   
1307      0              359309    8.0500  ...          0          0   
1308      1                2668   22.3583  ...          1          0   

      Embarked_S  Ticket_value  Ticket_number  Ticket_value_encoded  \
0              1           A/5        21171.0                     1   
1              0            PC        17599.0                     2   
2              1      STON/O2.      3101282.0                     3   
3              1           NaN       113803.0                     4   
4              1           NaN       373450.0                     4   
...          ...           ...            ...                   ...   
1304           1          A.5.         3236.0                    25   
1305           0            PC        17758.0                     2   
1306           1    SOTON/O.Q.      3101262.0                    22   
1307           1           NaN       359309.0                     4   
1308           0           NaN         2668.0                     4   

      Ticket_final Cabin_encoded  Cabin_number  Cabin_combined  
0         211711.0           0.0             0               0  
1         175992.0           3.0            85            3085  
2       31012823.0           0.0             0               0  
3        1138034.0           3.0           123           30123  
4        3734504.0           0.0             0               0  
...            ...           ...           ...             ...  
1304      323625.0           0.0             0               0  
1305      177582.0           3.0           105           30105  
1306   310126222.0           0.0             0               0  
1307     3593094.0           0.0             0               0  
1308       26684.0           0.0             0               0  

[1309 rows x 24 columns]

In [ ]:
# Procurando valores nulos
dataset_cleaned.isnull().sum()

PassengerId               0
Survived                418
Pclass                    0
Name                      0
Sex                       0
Age                       0
SibSp                     0
Parch                     0
Ticket                    0
Fare                      0
Cabin                     0
Embarked                  0
Sex_female                0
Sex_male                  0
Embarked_C                0
Embarked_Q                0
Embarked_S                0
Ticket_value            957
Ticket_number             5
Ticket_value_encoded      0
Ticket_final              5
Cabin_encoded             0
Cabin_number              0
Cabin_combined            0
dtype: int64

In [ ]:
# Substitui os valores ausentes em "Ticket_number" e "Ticket_final" pela média
dataset_cleaned['Ticket_number'].fillna(int(dataset_cleaned['Ticket_number'].describe()[1]), inplace=True)
dataset_cleaned['Ticket_final'].fillna(int(dataset_cleaned['Ticket_final'].describe()[1]), inplace=True)

# Procurando valores nulos
dataset_cleaned.isnull().sum()

PassengerId               0
Survived                418
Pclass                    0
Name                      0
Sex                       0
Age                       0
SibSp                     0
Parch                     0
Ticket                    0
Fare                      0
Cabin                     0
Embarked                  0
Sex_female                0
Sex_male                  0
Embarked_C                0
Embarked_Q                0
Embarked_S                0
Ticket_value            957
Ticket_number             0
Ticket_value_encoded      0
Ticket_final              0
Cabin_encoded             0
Cabin_number              0
Cabin_combined            0
dtype: int64

Vamos agora criar uma cópia desse dataset para manter a integridade dos dados originais, e posteriormente excluir colunas desnecessárias.

In [ ]:
# Lista de colunas numéricas
colunas_numericas = dataset_cleaned.select_dtypes(include=[np.number]).columns.tolist()

# Lista de colunas não numéricas
colunas_nao_numericas = dataset_cleaned.select_dtypes(exclude=[np.number]).columns.tolist()

print("Colunas Numéricas:")
print(colunas_numericas)

print("\nColunas Não Numéricas:")
print(colunas_nao_numericas)

Colunas Numéricas:
['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Sex_female', 'Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S', 'Ticket_number', 'Ticket_value_encoded', 'Ticket_final', 'Cabin_encoded', 'Cabin_number', 'Cabin_combined']

Colunas Não Numéricas:
['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked', 'Ticket_value']


Podemos então excluir as seguintes colunas: 'Name', 'Sex', 'Ticket', 'Cabin', 'Embarked', 'Ticket_value'

In [ ]:
# Lista das colunas a serem excluídas
columns_to_drop = colunas_nao_numericas

# Excluindo as colunas da lista
dataset_cleaned.drop(columns=columns_to_drop, inplace=True)

# Exibindo as primeiras linhas do DataFrame após a exclusão das colunas
dataset_cleaned

PassengerId  Survived  Pclass   Age  SibSp  Parch      Fare  Sex_female  \
0               1       0.0       3  22.0      1      0    7.2500           0   
1               2       1.0       1  38.0      1      0   71.2833           1   
2               3       1.0       3  26.0      0      0    7.9250           1   
3               4       1.0       1  35.0      1      0   53.1000           1   
4               5       0.0       3  35.0      0      0    8.0500           0   
...           ...       ...     ...   ...    ...    ...       ...         ...   
1304         1305       NaN       3  29.0      0      0    8.0500           0   
1305         1306       NaN       1  39.0      0      0  108.9000           1   
1306         1307       NaN       3  38.5      0      0    7.2500           0   
1307         1308       NaN       3  29.0      0      0    8.0500           0   
1308         1309       NaN       3  29.0      1      1   22.3583           0   

      Sex_male  Embarked_C  Embarked_Q  Embarked_S  Ticket_number  \
0            1           0           0           1        21171.0   
1            0           1           0           0        17599.0   
2            0           0           0           1      3101282.0   
3            0           0           0           1       113803.0   
4            1           0           0           1       373450.0   
...        ...         ...         ...         ...            ...   
1304         1           0           0           1         3236.0   
1305         0           1           0           0        17758.0   
1306         1           0           0           1      3101262.0   
1307         1           0           0           1       359309.0   
1308         1           1           0           0         2668.0   

      Ticket_value_encoded  Ticket_final  Cabin_encoded  Cabin_number  \
0                        1      211711.0            0.0             0   
1                        2      175992.0            3.0            85   
2                        3    31012823.0            0.0             0   
3                        4     1138034.0            3.0           123   
4                        4     3734504.0            0.0             0   
...                    ...           ...            ...           ...   
1304                    25      323625.0            0.0             0   
1305                     2      177582.0            3.0           105   
1306                    22   310126222.0            0.0             0   
1307                     4     3593094.0            0.0             0   
1308                     4       26684.0            0.0             0   

      Cabin_combined  
0                  0  
1               3085  
2                  0  
3              30123  
4                  0  
...              ...  
1304               0  
1305           30105  
1306               0  
1307               0  
1308               0  

[1309 rows x 18 columns]

In [ ]:
# Lista de colunas numéricas
colunas_numericas = dataset_cleaned.select_dtypes(include=[np.number]).columns.tolist()

# Lista de colunas não numéricas
colunas_nao_numericas = dataset_cleaned.select_dtypes(exclude=[np.number]).columns.tolist()

print("Colunas Numéricas:")
print(colunas_numericas)

print("\nColunas Não Numéricas:")
print(colunas_nao_numericas)

Colunas Numéricas:
['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'Sex_female', 'Sex_male', 'Embarked_C', 'Embarked_Q', 'Embarked_S', 'Ticket_number', 'Ticket_value_encoded', 'Ticket_final', 'Cabin_encoded', 'Cabin_number', 'Cabin_combined']

Colunas Não Numéricas:
[]


Ok, agora todas as colunas são numéricas e podem ser utilizadas adequadamente para o aprendizado de máquina. Vamos separar novamente os sets de treino e teste. Separando o dataset_cleaned em dataset_training_cleaned e dataset_test_cleaned, as linhas que contem valor NaN na coluna Survived serão do dataset_test_cleaned, o restante pertencerá ao dataset_training_cleaned

In [ ]:
# Criando dataset_training_cleaned excluindo as linhas com NaN em "Survived"
dataset_training_cleaned = dataset_cleaned.dropna(subset=['Survived'])

# Criando dataset_test_cleaned com linhas que contêm NaN em "Survived"
dataset_test_cleaned = dataset_cleaned[dataset_cleaned['Survived'].isna()]

In [ ]:
dataset_test_cleaned

PassengerId  Survived  Pclass   Age  SibSp  Parch      Fare  Sex_female  \
891           892       NaN       3  34.5      0      0    7.8292           0   
892           893       NaN       3  47.0      1      0    7.0000           1   
893           894       NaN       2  62.0      0      0    9.6875           0   
894           895       NaN       3  27.0      0      0    8.6625           0   
895           896       NaN       3  22.0      1      1   12.2875           1   
...           ...       ...     ...   ...    ...    ...       ...         ...   
1304         1305       NaN       3  29.0      0      0    8.0500           0   
1305         1306       NaN       1  39.0      0      0  108.9000           1   
1306         1307       NaN       3  38.5      0      0    7.2500           0   
1307         1308       NaN       3  29.0      0      0    8.0500           0   
1308         1309       NaN       3  29.0      1      1   22.3583           0   

      Sex_male  Embarked_C  Embarked_Q  Embarked_S  Ticket_number  \
891          1           0           1           0       330911.0   
892          0           0           0           1       363272.0   
893          1           0           1           0       240276.0   
894          1           0           0           1       315154.0   
895          0           0           0           1      3101298.0   
...        ...         ...         ...         ...            ...   
1304         1           0           0           1         3236.0   
1305         0           1           0           0        17758.0   
1306         1           0           0           1      3101262.0   
1307         1           0           0           1       359309.0   
1308         1           1           0           0         2668.0   

      Ticket_value_encoded  Ticket_final  Cabin_encoded  Cabin_number  \
891                      4     3309114.0            0.0             0   
892                      4     3632724.0            0.0             0   
893                      4     2402764.0            0.0             0   
894                      4     3151544.0            0.0             0   
895                      4    31012984.0            0.0             0   
...                    ...           ...            ...           ...   
1304                    25      323625.0            0.0             0   
1305                     2      177582.0            3.0           105   
1306                    22   310126222.0            0.0             0   
1307                     4     3593094.0            0.0             0   
1308                     4       26684.0            0.0             0   

      Cabin_combined  
891                0  
892                0  
893                0  
894                0  
895                0  
...              ...  
1304               0  
1305           30105  
1306               0  
1307               0  
1308               0  

[418 rows x 18 columns]

In [ ]:
dataset_training_cleaned

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_female  \
0              1       0.0       3  22.0      1      0   7.2500           0   
1              2       1.0       1  38.0      1      0  71.2833           1   
2              3       1.0       3  26.0      0      0   7.9250           1   
3              4       1.0       1  35.0      1      0  53.1000           1   
4              5       0.0       3  35.0      0      0   8.0500           0   
..           ...       ...     ...   ...    ...    ...      ...         ...   
886          887       0.0       2  27.0      0      0  13.0000           0   
887          888       1.0       1  19.0      0      0  30.0000           1   
888          889       0.0       3  29.0      1      2  23.4500           1   
889          890       1.0       1  26.0      0      0  30.0000           0   
890          891       0.0       3  32.0      0      0   7.7500           0   

     Sex_male  Embarked_C  Embarked_Q  Embarked_S  Ticket_number  \
0           1           0           0           1        21171.0   
1           0           1           0           0        17599.0   
2           0           0           0           1      3101282.0   
3           0           0           0           1       113803.0   
4           1           0           0           1       373450.0   
..        ...         ...         ...         ...            ...   
886         1           0           0           1       211536.0   
887         0           0           0           1       112053.0   
888         0           0           0           1         6607.0   
889         1           1           0           0       111369.0   
890         1           0           1           0       370376.0   

     Ticket_value_encoded  Ticket_final  Cabin_encoded  Cabin_number  \
0                       1      211711.0            0.0             0   
1                       2      175992.0            3.0            85   
2                       3    31012823.0            0.0             0   
3                       4     1138034.0            3.0           123   
4                       4     3734504.0            0.0             0   
..                    ...           ...            ...           ...   
886                     4     2115364.0            0.0             0   
887                     4     1120534.0            2.0            42   
888                    16      660716.0            0.0             0   
889                     4     1113694.0            3.0           148   
890                     4     3703764.0            0.0             0   

     Cabin_combined  
0                 0  
1              3085  
2                 0  
3             30123  
4                 0  
..              ...  
886               0  
887            2042  
888               0  
889           30148  
890               0  

[891 rows x 18 columns]

Salvando os dataframes limpos em arquivo .csv

In [ ]:
# Salvando dataset_cleaned em um arquivo .csv
dataset_cleaned.to_csv('dataset_cleaned.csv', index=False)

# Salvando dataset_training_cleaned em um arquivo .csv
dataset_training_cleaned.to_csv('dataset_training_cleaned.csv', index=False)

# Salvando dataset_test_cleaned em um arquivo .csv
dataset_test_cleaned.to_csv('dataset_test_cleaned.csv', index=False)

In [ ]:
# Checando se ainda há algum valor null
dataset_cleaned.isnull().sum()

PassengerId               0
Survived                418
Pclass                    0
Age                       0
SibSp                     0
Parch                     0
Fare                      0
Sex_female                0
Sex_male                  0
Embarked_C                0
Embarked_Q                0
Embarked_S                0
Ticket_number             0
Ticket_value_encoded      0
Ticket_final              0
Cabin_encoded             0
Cabin_number              0
Cabin_combined            0
dtype: int64

# O que falta?

-Normalizar os dados não compensa muito nesse caso, pois o enviesamento é o ideal para a tomada de decisão das árvores

-Mais alguma coisa para tratar?

-Olhar o trello

-Olhar a documentação

-Revisar todas as células e códigos

-Prazo da meta desse sprint: 31/08/2023